Initial LOD proposal

meshula opened the pull request:
### Description of Proposal

This proposal defines an API schema for managing Level of Detail (LOD) in USD stages. The LOD API enables creators and renderers to efficiently switch between different detail levels of assets based on configurable criteria like distance, screen size, or explicit selection. This approach allows optimizing rendering performance while maintaining visual quality where needed.

This proposal is a modernization of the original Ubisoft LoD proposal which is included in the proposal as supporting documentation. Many thanks to the original authors for that work.

This proposal is preliminary, for community discussion and requirements gathering, and not expected to land in its present form. Please use this PR as a place to collate discussion and feedback. It would be great to learn of alternative approaches, modern LoD systems, how LoD works in _your_ engine, and so on.

[<!--Replace the link here with a link to your own proposal markdown file -->
[Link to Rendered Proposal](https://github.com/PixarAnimationStudios/OpenUSD-proposals/blob/main/README.md)
](https://github.com/PixarAnimationStudios/OpenUSD-proposals/tree/lod/proposals/level-of-detail)


### Contributing


- [X] I agree to and accept the [Supplemental Terms](https://graphics.pixar.com/usd/release/contributing_supplemental.html).

meshula committed:
Initial LOD proposal

meshula committed:
Merge branch 'main' into lod

meshula committed:
Update lod proposal incorporating feedback to date

meshula commented:
The proposal has been updated to reflect feedback to date, and now includes a definition of Heterogeneous Hierarchical Level of Detail.

> How could we do paired LOD? E.g if you have crowds, you might have mesh+skeleton+material that change together versus independently.

@dgovil The proposal has been updated to concretely cover that case.

meshula committed:
clarify canonical vs advisory usages

meshula committed:
Revise proposal to correctly prepend LODHeuristic

meshula commented:
Proposal updated with most recent feedback taken into account.

meshula committed:
Merge branch 'main' into lod

meshula committed:
Update schema with correct syntax

dgovil commented:
So just some feedback items:

* I think the API names are a little too generic. ConfigurationAPI and LevelAPI aren’t clear by themselves and would preclude anyone else having something called configuration api. Which I can see people wanting that in the future for other domains too

* same with selectionMode and targetFramerate, unless that gets namespaced when applied? 

* I think we’re still missing how to tell USD not to render a specific hierarchy until it’s activated by an LOD. The example scene I think would have people rendering the city hierarchy in its entirety. 


But from a “what needs to be defined” perspective I think it makes sense.

spiffmon commented:
The schema declaration can be configured to preserve just the "Lod" part of the prefix for the schema name/identifier, while still allowing the C++ classes to have the full "UsdLod" prefix, [as we do for the schemas in UsdRender](https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdRender/schema.usda#L131-L140).

dgovil commented:
Yeah I think that would be preferable. That way it's LODConfigurationAPI which feels much nicer in the file.

meshula committed:
Revise per latest feedback round

o2b3d committed:
Updated version of the level-of-detail proposal

The proposal has been redone, guided by an implementation in code that
clarified issues in the previous version.

README.md contains the new proposal details.

There are 5 new schemas described in the proposal and implemented in
schema.usda: LODRootAPI, LODHeuristic, LODDistanceHeuristic,
LODScreenSizeHeuristic, and LODOverrideAPI.

api-desgin.md, example.usda, and lod-diagram.md have been updated.

A note has been added to use-cases.md indicating that it is out of date.
It will be fully updated in a future commit.

o2b3d commented on proposals/level-of-detail/example.usda:
```diff
@@ -7,69 +7,44 @@ def Xform "ExampleModel" (
+
+            # High detail geometry hierarchy goes here. It is invisible by
+            # default because it is not the default level of detail (see
+            # lod.default.index above).  Legacy renderers will not render it.
```
Typo. Attribute name is "lod:default:index" with colons rather than dots.

dsyu-pixar committed:
Merge branch 'main' into lod
